brick/dockerlib.py

```diff
@@ -119,13 +121,55 @@ def docker_build(
             logs = [cmd]
             logger.debug(cmd)
 
+            # State machine keeping track of step
+            step_id = None
+            step_command = None
+            step_is_cacheable = None  # Some steps can't be cached
+
             while p.poll() is None:
                 line = p.stdout.readline()
                 if line != "":
                     line = line.rstrip("\n")
                     logs.append(line)
                     logger.debug(line)
 
+                    # A line is typically "#9 [3/6] COPY ...."
+                    # Followed by either
+                    # #9 CACHED
+                    # or
+                    # #9 DONE 0.0s
+
+                    # Detect step id
+                    step_id_match = re.match(r"#(?P<id>\d+)", line)
+                    if step_id_match:
+                        step_id = step_id_match.group("id")
+                    else:
+                        # Reset
+                        step_id = None
+                        step_command = None
+                        step_is_cacheable = None
+
+                    # Extra step extended info
+                    step_match = re.match(
+                        r"#(?P<id>\d+) \[.*(?P<number>\d+)/\d+\] (?P<command>.*)", line
+                    )
+                    if step_match:
+                        assert step_id == step_match.group("id")
+                        step_command = step_match.group("command")
+                        # Steps of the form "#X [ Y/Z] ..." can be cached
+                        # However, "FROM" commands can't
+                        step_is_cacheable = not step_command.startswith("FROM ")
+                    cache_invalidated_match = line.startswith(f"#{step_id} DONE")
+                    if (
+                        line.startswith(f"#{step_id}")
+                        and step_is_cacheable
+                        and cache_invalidated_match
+                        and is_cached
+                    ):
+                        # Cache has been invalidated
+                        logger.info(f"Cache invalidated by {step_command}")
```

Comment:
I'm considering if this line should be a debug log instead of info. But as Brick is fairly verbose already, adding another info log shouldn't be a concern.